File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/novel/adapters/NovelPagerAdapter.java
package com.github.doomsdayrs.apps.shosetsu.ui.novel.adapters;

import android.util.Log;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;

import java.util.List;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 */
public class NovelPagerAdapter extends FragmentPagerAdapter {
    private final List<Fragment> fragments;
    private String[] titles = {"Info", "Chapters"};

    public NovelPagerAdapter(FragmentManager fm, List<Fragment> fragments) {
        super(fm);
        this.fragments = fragments;
    }

    //TODO with tracker use this instead the of the above
    public NovelPagerAdapter(FragmentManager fm, List<Fragment> fragments, boolean ignored) {
        super(fm);
        this.fragments = fragments;
        titles = new String[]{titles[0], titles[1], "Tracker"};
    }

    @Override
    public Fragment getItem(int i) {
        Log.d("SwapScreen", fragments.get(i).toString());
        return fragments.get(i);
    }

    @Override
    public int getCount() {
        return fragments.size();
    }

    @Nullable
    @Override
    public CharSequence getPageTitle(int position) {
        return titles[position];
    }
}

File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/novel/adapters/NovelPagerAdapter.kt
package com.github.doomsdayrs.apps.shosetsu.ui.novel.adapters

import android.util.Log
import android.widget.ArrayAdapter
import androidx.core.os.bundleOf
import androidx.viewpager.widget.ViewPager
import com.bluelinelabs.conductor.Controller
import com.bluelinelabs.conductor.Router
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.support.RouterPagerAdapter
import com.github.doomsdayrs.apps.shosetsu.R
import com.github.doomsdayrs.apps.shosetsu.common.consts.BundleKeys.BUNDLE_NOVEL_ID
import com.github.doomsdayrs.apps.shosetsu.common.ext.context
import com.github.doomsdayrs.apps.shosetsu.common.ext.getNovelID
import com.github.doomsdayrs.apps.shosetsu.common.ext.logID
import com.github.doomsdayrs.apps.shosetsu.ui.novel.NovelController
import com.github.doomsdayrs.apps.shosetsu.ui.novel.pages.NovelChaptersController
import com.github.doomsdayrs.apps.shosetsu.ui.novel.pages.NovelInfoController
import com.github.doomsdayrs.apps.shosetsu.view.base.FABView
import com.google.android.material.floatingactionbutton.FloatingActionButton

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Shosetsu
 * 9 / June / 2019
 */
class NovelPagerAdapter(private val novelController: NovelController)
	: RouterPagerAdapter(novelController) {
	companion object {
		private const val INFO_CONTROLLER = 0
		private const val CHAPTERS_CONTROLLER = 1
		private const val TRACK_CONTROLLER = 2
	}

	private val titles by lazy {
		ArrayAdapter(
				novelController.context!!,
				android.R.layout.simple_spinner_item,
				novelController.resources!!.getStringArray(R.array.novel_fragment_names)
		)
	}

	private val controllers: Array<Controller> = arrayOf(
			NovelInfoController(bundleOf(
					BUNDLE_NOVEL_ID to novelController.bundle.getNovelID()
			)),
			NovelChaptersController(bundleOf(
					BUNDLE_NOVEL_ID to novelController.bundle.getNovelID()
			))
	)


	override fun configureRouter(router: Router, position: Int) {
		if (!router.hasRootController()) {
			router.setRoot(RouterTransaction.with(controllers[position]))
		}
	}

	override fun getCount(): Int = controllers.size

	override fun getPageTitle(position: Int): CharSequence? = titles.getItem(position)

	inner class PageController(
			private val fab: FloatingActionButton,
	) : ViewPager.OnPageChangeListener {
		private var currentPosition = 0

		override fun onPageScrollStateChanged(state: Int) {
		}

		override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
		}

		override fun onPageSelected(position: Int) {
			if (currentPosition != position) {
				Log.d(logID(), "Current position is $currentPosition")
				Log.d(logID(), "TheNext position is $position")

				val currentController = controllers[currentPosition]
				if (currentController is FABView) {
					currentController.hideFAB(fab)
					currentController.resetFAB(fab)
				}

				val newController = controllers[position]
				if (newController is FABView) {
					newController.setFABIcon(fab)
					newController.manipulateFAB(fab)
					newController.showFAB(fab)
				}
				currentPosition = position
			}
		}
	}
}